src/createStore.ts

```diff
@@ -8,7 +12,36 @@ let createStoreAction = action('createStore', function createStoreAction(
     getRootStore().set(key, initialState);
 });
 
-export default function createStore<T>(key: string, initialState: T): () => T {
+export default function createStore<TState>(
+    key: string,
+    arg2: TState | LeafMutator<TState> | CombinedMutator<TState>
+): () => TState {
+    // Get the initial state (from the mutator, if necessary)
+    let mutator = getMutator(arg2);
+    let initialState = mutator ? mutator.getInitialValue() : arg2;
+
+    // Create the store under the root store
     createStoreAction(key, initialState);
-    return () => <T>getRootStore().get(key);
+    let getStore = () => <TState>getRootStore().get(key);
+
+    // If necessary, hook the mutator up to the dispatcher
+    if (mutator) {
+        subscribeAll(
+            wrapMutator(actionMessage => {
+                mutator.handleAction(getStore(), actionMessage, newState => {
+                    getRootStore().set(key, newState);
+                });
+            })
+        );
+    }
+
+    return getStore;
+}
+
+function getMutator<TState>(mutator: TState | LeafMutator<TState> | CombinedMutator<TState>) {
```

Comment:
Would it make more sense to implement a type guard method instead?
```typescript
function getMutator<TState>(mutator: TState | LeafMutator<TState> | CombinedMutator<TState>): mutator is LeafMutator<TState> | CombinedMutator<TState> {
  let typedMutator = <LeafMutator<TState> | CombinedMutator<TState>>mutator;
  return !!(typedMutator.handleAction && typedMutator.getInitialValue);
}

Reply:
I could, but the code here doesn't lend itself so well to the `if {...} else {...}` structure that a type guard generally follows.  IMHO assigning it to a new `mutator` variable makes the code clearer.

---
In reply to: [200791792](https://github.com/Microsoft/satcheljs/pull/100#discussion_r200791792) [](ancestors = 200791792)